python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py

```diff
@@ -604,3 +610,201 @@ def partitioned_asset(context: AssetExecutionContext) -> MaterializeResult:
 
     res = partitioned_asset(context)
     assert res.metadata["key"] == "red"  # pyright: ignore[reportAttributeAccessIssue]
+
+
+def test_materialize_result_value():
+    @asset
+    def asset_with_value():
+        return MaterializeResult(value="hello")
+
+    result = materialize([asset_with_value])
+
+    assert result.success
+    assert result.asset_value("asset_with_value") == "hello"
+
+
+def test_materialize_result_value_annotated_no_type():
+    @asset
+    def asset_with_value() -> MaterializeResult:
+        return MaterializeResult(value="hello")
+
+    result = materialize([asset_with_value])
+
+    assert result.success
+    assert result.asset_value("asset_with_value") == "hello"
+
+
+def test_materialize_result_value_annotated_explicit_type():
+    @asset
+    def asset_with_value() -> MaterializeResult[str]:
+        return MaterializeResult(value="hello")
+
+    result = materialize([asset_with_value])
+
+    assert result.success
+    assert result.asset_value("asset_with_value") == "hello"
+
+
+def test_materialize_result_value_annotated_incorrect_type():
+    @asset
+    def asset_with_value() -> MaterializeResult[int]:
+        return MaterializeResult(value="hello")  # type: ignore
+
+    with pytest.raises(DagsterTypeCheckDidNotPass):
+        materialize([asset_with_value])
+
+
+def test_materialize_result_value_annotated_no_value():
+    @asset
+    def asset_with_value() -> MaterializeResult[int]:
+        return MaterializeResult()  # type: ignore
+
+    with pytest.raises(DagsterTypeCheckDidNotPass):
+        materialize([asset_with_value])
```

Comment:
most of these tests were copied over from the original PR, but these are net-new (suggested by @alangenfeld )